certbot-postfix/certbot_postfix/util.py

```diff
@@ -0,0 +1,285 @@
+    :raises subprocess.CalledProcessError: if the command fails
+
+    """
+    for keyword in ('stdout', 'stderr', 'universal_newlines',):
+        if keyword in kwargs:
+            raise ValueError(
+                keyword + ' argument not allowed, it will be overridden.')
+
+    kwargs['stdout'] = subprocess.PIPE
+    kwargs['stderr'] = subprocess.PIPE
+    kwargs['universal_newlines'] = True
+
+    process = subprocess.Popen(*args, **kwargs)
+    output, err = process.communicate()
+    retcode = process.poll()
+    if retcode:
+        cmd = kwargs.get('args')
+        if cmd is None:
+            cmd = args[0]
+        logger.debug(
+            "'%s' exited with %d. stdout output was:\n%s\nstderr output was:\n%s",
+            cmd, retcode, output, err)
+        raise subprocess.CalledProcessError(retcode, cmd)
+    return (output, err)
+
+
+def verify_exe_exists(exe, message=None):
+    """Ensures an executable with the given name is available.
+
+    If an executable isn't found for the given path or name, extra
+    directories are added to the user's PATH to help find system
+    utilities that may not be available in the default cron PATH.
+
+    :param str exe: executable path or name
+    :param str message: Error message to print.
+
+    :raises .NoInstallationError: when the executable isn't found
+
+    """
+    if message is None:
+        message = "Cannot find executable '{0}'.".format(exe)
+    if not (certbot_util.exe_exists(exe) or plugins_util.path_surgery(exe)):
+        raise errors.NoInstallationError(message)
+
+def report_master_overrides(name, overrides, acceptable_overrides=None):
+    """If the value for a parameter |name| is overridden by other services,
+    report a warning to notify the user.
+
+    :param str name: The name of the parameter that is being overridden.
+    :param list overrides: The values that other services are setting for |name|.
+        Each override is a tuple: (service name, value)
+    :param tuple acceptable_overrides: Override values that are acceptable. For instance, if
+        another service is overriding our parameter with a more secure option, we don't have
+        to warn. If this is set to None, errors are raised for *any* overrides of `name`!
+    """
+    error_string = ""
+    for override in overrides:
+        service, value = override
+        # If this override is acceptable:
+        if acceptable_overrides is not None and \
+            is_acceptable_value(name, value, acceptable_overrides):
+            continue
+        error_string += "  {0}: {1}\n".format(service, value)
+    if error_string:
+        raise errors.PluginError("{0} is overridden with less secure options by the "
+             "following services in master.cf:\n".format(name) + error_string)
+
+def is_acceptable_value(parameter, value, acceptable):
+    """ Returns whether the `value` for this `parameter` is acceptable,
+    given a string or tuple `acceptable`
+
```

Comment:
If `tls_protocols` is in `parameter`, `acceptable` isn't used but it's still required to be set.

Maybe we could refactor things a bit and move this `tls_protocols` check outside of this function, however, I think we should at least add a comment describing this behavior.

Reply:
Yeah, I agree it's weird. I thought about changing the variable mappings in `constants.py` so that they explicitly define acceptability (maybe as a `callable(str) => bool` ) and default value separately for each parameter, since the tuples are being used for both. This would involve a larger refactor.

So I just added a comment, here and for `report_master_overrides`.